zebrad/src/components/inbound.rs

```diff
@@ -155,9 +155,17 @@ impl Service<zn::Request> for Inbound {
                 debug!("ignoring unimplemented request");
                 async { Ok(zn::Response::Nil) }.boxed()
             }
-            zn::Request::FindBlocks { .. } => {
-                debug!("ignoring unimplemented request");
-                async { Ok(zn::Response::Nil) }.boxed()
+            zn::Request::FindBlocks { known_blocks, stop } => {
+                let request = zs::Request::GetHashes(known_blocks, stop);
+                let mut state = self.state.clone();
+                async move {
+                    let fut = state.ready_and().await?.call(request).await?;
+                    match fut {
+                        zs::Response::Hashes(Some(hashes)) => Ok(zn::Response::BlockHashes(hashes)),
+                        _ => Ok(zn::Response::Nil),
+                    }
+                }
+                .boxed()
```

Comment:
The `poll_ready` implementation for the inbound service already calls `ready_and`, we don't need to call it again here. Instead we should probably just return the future produced by the state's call fn and just transform that with combinators.

I also tweaked the `_` branch of the match to instead panic since that should not be possible to trigger unless we have a bug.

```suggestion
                state.call(request).map_ok(|response| match fut {
                        zs::Response::Hashes(Some(hashes)) => Ok(zn::Response::BlockHashes(hashes)),
                        _ => unreachable!("zebra-state should always respond to a `GetHashes` request with a `Hashes` response"),
                    }).boxed()
```

Reply:
> I also tweaked the `_` branch of the match to instead panic since that should not be possible to trigger unless we have a bug.

We can't panic here, because the `zn::Request::FindBlocks` request is untrusted data - it comes from peers. A peer can provide any hash, including hashes that never existed in any chain.

If that happens, the `zs::Response::Hashes` will be empty, and the network should return `zn::Response::Nil`.

Panicking would let remote nodes crash Zebra.

Reply:
Then we should handle the `zs::Response::Hashes(None)` with a `Nil` reply and panic on all other responses.

Reply:
If i do this i get a crash:

```
The application panicked (crashed).
Message:  buffer full; poll_ready must be called first
Location: /home/oxarbitrage/.cargo/git/checkouts/tower-b098c32cf5a1bcca/d4d1c67/tower/src/buffer/service.rs:129

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  
   0: zebrad::components::inbound::call with req=FindBlocks { known_blocks: [block::Hash("00000003d45d644c0bb4c12cdda2f1c697ec21840ab806fb119ea93350d7b21c"), block::Hash("00000002497348fbe42d3d1c134c18bbd065d1deaecb3935f26c5de3bbd72bbe"), block::Hash("00000002ec4c5c54de9c4116ec715e86b5c5f103b79ddd56b3774dd50b60131f"), block::Hash("00000000d888ecf9e8513383d885837b3cf206b1df1761d6e65af28b45c7102b"), block::Hash("000000146ba76f69ec6dc23bd74adec6870d228831591ea3c203ee46540ae7a9"), block::Hash("000000142f9acfaa55229bbae9436774e3a6b037bf3a8fe95e6aee7762cbb13c"), block::Hash("00000007d6805fb388db8f72d96d8ca9ef1b37b7be121c25ba9250dedebdc341"), block::Hash("0000000afa664031e2c4cec3e5f728bfa353f97c4dcec9f900b4f3503b929782"), block::Hash("0000000d4d229617e388f17ce1420c0ca0e2a75a1023b7d6e733ae7f12512d73")], stop: None }
      at zebrad/src/components/inbound.rs:116
   1: zebra_network::peer::connection::handle_message_as_request with msg=GetBlocks { known_blocks: [block::Hash("00000003d45d644c0bb4c12cdda2f1c697ec21840ab806fb119ea93350d7b21c"), block::Hash("00000002497348fbe42d3d1c134c18bbd065d1deaecb3935f26c5de3bbd72bbe"), block::Hash("00000002ec4c5c54de9c4116ec715e86b5c5f103b79ddd56b3774dd50b60131f"), block::Hash("00000000d888ecf9e8513383d885837b3cf206b1df1761d6e65af28b45c7102b"), block::Hash("000000146ba76f69ec6dc23bd74adec6870d228831591ea3c203ee46540ae7a9"), block::Hash("000000142f9acfaa55229bbae9436774e3a6b037bf3a8fe95e6aee7762cbb13c"), block::Hash("00000007d6805fb388db8f72d96d8ca9ef1b37b7be121c25ba9250dedebdc341"), block::Hash("0000000afa664031e2c4cec3e5f728bfa353f97c4dcec9f900b4f3503b929782"), block::Hash("0000000d4d229617e388f17ce1420c0ca0e2a75a1023b7d6e733ae7f12512d73")], stop: None }
      at zebra-network/src/peer/connection.rs:564
   2: zebra_network::peer::handshake::peer with addr=127.0.0.1:36940
      at zebra-network/src/peer/handshake.rs:195
```

Reply:
solved by https://github.com/ZcashFoundation/zebra/pull/1347/commits/20bc94d17ee3a0b9f7fd1288800b443aa655109d